File: backend/app/agents/ots_agent.py
"""OTSAgent for OT ingestion from TELCOS API."""

import logging
from datetime import datetime
from backend.app.agents.base_agent import BaseAgent
from backend.app.agents.state import AgentState
from backend.app.models.ot import OrdenTrabajo, OTStatus, ProjectType
from backend.app.services.telcos_api_client import TelcosApiClient
from backend.app.utils.geo_utils import validate_coordinates

logger = logging.getLogger(__name__)


class OTSAgent(BaseAgent):
    """Ingests OTs from TELCOS API and persists to database."""

    async def execute(self, state: AgentState) -> AgentState:
        """Sync OTs from TELCOS API."""
        try:
            logger.info("OTSAgent: Starting OT sync")
            
            # Get OTs from TELCOS API
            telcos_client = TelcosApiClient()
            telcos_ots = await telcos_client.get_ots()
            logger.info(f"OTSAgent: Fetched {len(telcos_ots)} OTs from TELCOS")
            
            created_ots = []
            errors = []
            
            # Process each OT
            for telcos_ot in telcos_ots:
                try:
                    # Check if OT already exists
                    existing = self.db_session.query(OrdenTrabajo).filter(
                        OrdenTrabajo.external_id == telcos_ot.external_id
                    ).first()
                    
                    if existing:
                        logger.debug(f"OT {telcos_ot.external_id} already exists, skipping")
                        continue
                    
                    # Validate coordinates
                    error_geo = False
                    if telcos_ot.lat is not None and telcos_ot.long is not None:
                        if not validate_coordinates(telcos_ot.lat, telcos_ot.long):
                            error_geo = True
                            logger.warning(f"Invalid coordinates for {telcos_ot.external_id}")
                    
                    # Create new OT
                    new_ot = OrdenTrabajo(
                        external_id=telcos_ot.external_id,
                        status=OTStatus.PREPLANIFICADA,
                        project_type=telcos_ot.project_type,
                        lat=telcos_ot.lat,
                        long=telcos_ot.long,
                        cliente_id=telcos_ot.cliente_id,
                        login_id=telcos_ot.login_id,
                        error_geo=error_geo,
                        created_at=datetime.utcnow(),
                        updated_at=datetime.utcnow(),
                    )
                    
                    self.db_session.add(new_ot)
                    created_ots.append(str(new_ot.id))
                    
                except Exception as e:
                    error_msg = f"Error processing OT {getattr(telcos_ot, 'external_id', 'unknown')}: {str(e)}"
                    logger.error(error_msg)
                    errors.append(error_msg)
            
            # Commit changes
            self.db_session.commit()
            
            # Log action
            self.log_action(
                action="SYNC_OTS",
                resultado="success" if not errors else "partial_success",
                metadata={
                    "created_count": len(created_ots),
                    "error_count": len(errors),
                    "created_ot_ids": created_ots,
                }
            )
            
            logger.info(f"OTSAgent: Sync complete - {len(created_ots)} created, {len(errors)} errors")
            
            # Update state
            state["agent_history"].append(self.agent_name)
            state["result"] = {
                "created_ots": created_ots,
                "created_count": len(created_ots),
                "error_count": len(errors),
                "errors": errors,
            }
            state["should_continue"] = True
            
            return state
            
        except Exception as e:
            logger.error(f"OTSAgent error: {str(e)}", exc_info=True)
            self.db_session.rollback()
            return self._update_state(state, error=str(e), should_continue=False)

File: backend/src/agents/OTSAgent.ts
import BaseAgent, { AgentResult } from './BaseAgent';
import { createApiService } from '../services/ApiService';
import { prisma } from '../lib/prisma';
import { logger } from '../utils/logger';

export interface OTSInput {
  source: 'telcos_api' | 'manual';
  data?: any;
}

export interface OTSError {
  otId: string;
  reason: string;
}

// Ecuador geographic bounds
const ECUADOR_BOUNDS = {
  minLat: -5,
  maxLat: 2,
  minLong: -82,
  maxLong: -75,
};

export class OTSAgent extends BaseAgent {
  private apiService = createApiService();

  constructor() {
    super('OTSAgent', 'Handles OT ingestion, coordinate validation, and database registration');
  }

  async execute(input: OTSInput): Promise<AgentResult> {
    try {
      const { source, data } = input;

      logger.info('OTSAgent executing', { source });

      if (source === 'telcos_api') {
        return await this.ingestFromTelcosAPI();
      } else if (source === 'manual') {
        return await this.ingestManual(data);
      } else {
        return {
          success: false,
          error: `Unknown source: ${source}`,
        };
      }
    } catch (error: any) {
      logger.error('OTSAgent execution failed', { error: error.message });

      return {
        success: false,
        error: `OTSAgent execution failed: ${error.message}`,
      };
    }
  }

  /**
   * Ingest OTs from TELCOS API via MockApiService
   */
  private async ingestFromTelcosAPI(): Promise<AgentResult> {
    const created: string[] = [];
    const errors: OTSError[] = [];
    let hasGeoError = false;

    try {
      logger.info('Fetching OTs from TELCOS API');
      const mockOTs = await this.apiService.getOTs();

      logger.info(`Retrieved ${mockOTs.length} OTs from TELCOS API`);

      // Process each OT
      for (const mockOT of mockOTs) {
        try {
          // Validate coordinates
          const coordValidation = this.validateCoordinates(
            mockOT.lat,
            mockOT.long
          );

          if (!coordValidation.valid) {
            hasGeoError = true;
            errors.push({
              otId: mockOT.external_id,
              reason: coordValidation.error!,
            });
            logger.warn('OT coordinate validation failed', {
              otId: mockOT.external_id,
              reason: coordValidation.error,
            });
          }

          // Create OT in database
          const ot = await prisma.oT.create({
            data: {
              externalId: mockOT.external_id,
              status: 'PREPLANIFICADA',
              projectType: mockOT.project_type as 'PUBLICO' | 'PRIVADO' | 'TERCERIZADO',
              clienteName: mockOT.cliente,
              loginId: mockOT.login,
              lat: mockOT.lat,
              long: mockOT.long,
              hasGeoError: !coordValidation.valid,
            },
          });

          created.push(ot.id);

          logger.info('OT created successfully', {
            otId: ot.id,
            externalId: ot.externalId,
            hasGeoError: ot.hasGeoError,
          });

          // Log to AgentLog if geo error
          if (!coordValidation.valid) {
            await prisma.agentLog.create({
              data: {
                otId: ot.id,
                agentName: 'OTSAgent',
                accion: 'COORDINATE_VALIDATION',
                resultado: 'Failed',
                errorMessage: coordValidation.error,
              },
            });
          }
        } catch (error: any) {
          logger.error('Error creating OT', {
            externalId: mockOT.external_id,
            error: error.message,
          });

          errors.push({
            otId: mockOT.external_id,
            reason: `Database error: ${error.message}`,
          });
        }
      }

      // Log summary
      await prisma.agentLog.create({
        data: {
          agentName: 'OTSAgent',
          accion: 'OT_INGESTION',
          resultado: `Ingested ${created.length} OTs, ${errors.length} errors`,
          rawLlmResponse: JSON.stringify({
            created: created.length,
            errors: errors.length,
            hasGeoError,
          }),
        },
      });

      logger.info('TELCOS API ingestion completed', {
        created: created.length,
        errors: errors.length,
        hasGeoError,
      });

      return {
        success: true,
        data: {
          created: created.length,
          errors,
          source: 'telcos_api',
        },
        llmResponse: JSON.stringify({
          created: created.length,
          errors: errors.length,
          hasGeoError,
        }),
        // If geo errors, route to CommunicationAgent for PM notification
        nextAgent: hasGeoError ? 'CommunicationAgent' : undefined,
      };
    } catch (error: any) {
      logger.error('TELCOS API ingestion failed', { error: error.message });

      return {
        success: false,
        error: `TELCOS API ingestion failed: ${error.message}`,
        data: {
          created: created.length,
          errors,
        },
      };
    }
  }

  /**
   * Ingest manually provided OT data
   */
  private async ingestManual(data: any): Promise<AgentResult> {
    const created: string[] = [];
    const errors: OTSError[] = [];
    let hasGeoError = false;

    if (!data || !Array.isArray(data)) {
      return {
        success: false,
        error: 'Manual ingestion requires array of OT data',
      };
    }

    try {
      logger.info('Ingesting OTs from manual data', { count: data.length });

      for (const otData of data) {
        try {
          // Validate required fields
          if (!otData.externalId || !otData.projectType) {
            errors.push({
              otId: otData.externalId || 'unknown',
              reason: 'Missing required fields: externalId, projectType',
            });
            continue;
          }

          // Validate coordinates
          const coordValidation = this.validateCoordinates(
            otData.lat,
            otData.long
          );

          if (!coordValidation.valid) {
            hasGeoError = true;
            errors.push({
              otId: otData.externalId,
              reason: coordValidation.error!,
            });
          }

          // Create OT
          const ot = await prisma.oT.create({
            data: {
              externalId: otData.externalId,
              status: otData.status || 'PREPLANIFICADA',
              projectType: otData.projectType,
              clienteName: otData.clienteName,
              loginId: otData.loginId,
              lat: otData.lat,
              long: otData.long,
              hasGeoError: !coordValidation.valid,
            },
          });

          created.push(ot.id);

          logger.info('Manual OT created', {
            otId: ot.id,
            externalId: ot.externalId,
          });
        } catch (error: any) {
          logger.error('Error creating manual OT', {
            externalId: otData.externalId,
            error: error.message,
          });

          errors.push({
            otId: otData.externalId || 'unknown',
            reason: error.message,
          });
        }
      }

      // Log summary
      await prisma.agentLog.create({
        data: {
          agentName: 'OTSAgent',
          accion: 'OT_INGESTION',
          resultado: `Ingested ${created.length} manual OTs, ${errors.length} errors`,
          rawLlmResponse: JSON.stringify({
            created: created.length,
            errors: errors.length,
            hasGeoError,
          }),
        },
      });

      return {
        success: true,
        data: {
          created: created.length,
          errors,
          source: 'manual',
        },
        llmResponse: JSON.stringify({
          created: created.length,
          errors: errors.length,
          hasGeoError,
        }),
        nextAgent: hasGeoError ? 'CommunicationAgent' : undefined,
      };
    } catch (error: any) {
      logger.error('Manual ingestion failed', { error: error.message });

      return {
        success: false,
        error: `Manual ingestion failed: ${error.message}`,
        data: {
          created: created.length,
          errors,
        },
      };
    }
  }

  /**
   * Validate OT coordinates are within Ecuador bounds
   */
  private validateCoordinates(
    lat: number | null | undefined,
    long: number | null | undefined
  ): { valid: boolean; error?: string } {
    // Check if coordinates are provided
    if (lat === null || lat === undefined || long === null || long === undefined) {
      return {
        valid: false,
        error: 'Missing coordinates (lat/long)',
      };
    }

    // Check if coordinates are numbers
    if (typeof lat !== 'number' || typeof long !== 'number') {
      return {
        valid: false,
        error: 'Invalid coordinate format (must be numbers)',
      };
    }

    // Check Ecuador bounds
    if (
      lat < ECUADOR_BOUNDS.minLat ||
      lat > ECUADOR_BOUNDS.maxLat ||
      long < ECUADOR_BOUNDS.minLong ||
      long > ECUADOR_BOUNDS.maxLong
    ) {
      return {
        valid: false,
        error: `Coordinates outside Ecuador bounds. Lat: ${lat} (${ECUADOR_BOUNDS.minLat} to ${ECUADOR_BOUNDS.maxLat}), Long: ${long} (${ECUADOR_BOUNDS.minLong} to ${ECUADOR_BOUNDS.maxLong})`,
      };
    }

    return { valid: true };
  }
}

export default OTSAgent;

